File: src/lib/llm.ts
import { isTauri, trackedInvoke } from "./tauri"

export const DEFAULT_MODEL = "deepseek-v4-flash"
export const DEFAULT_ENDPOINT = "https://api.deepseek.com/chat/completions"
const REQUEST_TIMEOUT_MS = 120_000
const MAX_CONTENT_CHARS = 50_000

export interface GenerateRequest {
  prompt: string
  apiKey?: string
  model?: string
  endpoint?: string
  signal?: AbortSignal
}

interface ChatCompletionResponse {
  choices?: Array<{
    message?: {
      content?: string
    }
  }>
}

function isTimeoutError(err: unknown): boolean {
  return err instanceof Error && err.name === "TimeoutError"
}

function isAbortError(err: unknown): boolean {
  return err instanceof Error && err.name === "AbortError"
}

/**
 * 合并「外部取消信号 + 超时」的 AbortSignal。不用 AbortSignal.timeout：
 * 旧 WebView2/浏览器未实现时抛 TypeError，会被下面的 catch 误判为网络错误。
 * 用 AbortController + setTimeout 兜底，兼容面更宽。
 */
function createAbortSignal(
  external?: AbortSignal,
): { signal: AbortSignal; cleanup: () => void } {
  const controller = new AbortController()
  const timeoutId = setTimeout(() => {
    controller.abort(
      new DOMException("The request timed out", "TimeoutError"),
    )
  }, REQUEST_TIMEOUT_MS)

  const onExternalAbort = () => controller.abort()
  if (external) {
    if (external.aborted) {
      controller.abort()
    } else {
      external.addEventListener("abort", onExternalAbort, { once: true })
    }
  }

  return {
    signal: controller.signal,
    cleanup: () => {
      clearTimeout(timeoutId)
      external?.removeEventListener("abort", onExternalAbort)
    },
  }
}

export async function generateDocument(request: GenerateRequest): Promise<string> {
  const model = request.model ?? DEFAULT_MODEL
  const endpoint = request.endpoint ?? DEFAULT_ENDPOINT

  if (isTauri()) {
    // invoke 不能序列化 AbortSignal；取消在 Rust 侧由 120s 超时兜底，
    // 前端靠 mounted 守卫保证关闭后不写 store。这里只做已中止的前置短路。
    if (request.signal?.aborted) throw "LLM_ABORTED"
    return trackedInvoke<string>("generate_document", {
      prompt: request.prompt,
      apiKey: request.apiKey,
      model,
      endpoint,
    })
  }

  // 浏览器模式:key 会暴露在前端,仅 dev 用。生产走 Tauri 侧。
  const apiKey = request.apiKey
  if (!apiKey || apiKey.trim().length === 0) {
    throw "LLM_NO_API_KEY"
  }

  const { signal, cleanup } = createAbortSignal(request.signal)

  let response: Response
  try {
    response = await fetch(endpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify({
        model,
        messages: [
          {
            role: "system",
            content:
              "You are a professional government document writer. Respond only with a valid JSON object.",
          },
          { role: "user", content: request.prompt },
        ],
        temperature: 0.3,
        response_format: { type: "json_object" },
      }),
      signal,
    })
  } catch (err) {
    cleanup()
    if (isAbortError(err)) throw "LLM_ABORTED"
    if (isTimeoutError(err)) throw "LLM_TIMEOUT"
    throw "LLM_NETWORK_ERROR"
  }

  if (!response.ok) {
    cleanup()
    throw `LLM_HTTP_${response.status}`
  }

  let data: ChatCompletionResponse
  try {
    data = (await response.json()) as ChatCompletionResponse
  } catch (err) {
    cleanup()
    if (isTimeoutError(err)) throw "LLM_TIMEOUT"
    if (isAbortError(err)) throw "LLM_ABORTED"
    throw "LLM_INVALID_RESPONSE"
  }
  cleanup()

  const content = data.choices?.[0]?.message?.content
  if (!content || content.trim().length === 0) {
    throw "LLM_INVALID_RESPONSE"
  }
  if (content.length > MAX_CONTENT_CHARS) {
    throw "LLM_INVALID_RESPONSE"
  }

  return content
}

File: src-tauri/src/llm.rs
use std::time::Duration;

use serde::Deserialize;

const DEFAULT_ENDPOINT: &str = "https://api.deepseek.com/chat/completions";
const DEFAULT_MODEL: &str = "deepseek-v4-flash";
const REQUEST_TIMEOUT: Duration = Duration::from_secs(120);
const API_KEY_ENV_VAR: &str = "EZDOC_API_KEY";

#[derive(Debug)]
enum LlmError {
    NoApiKey,
    Timeout,
    Network,
    Http(u16),
    InvalidResponse,
}

impl LlmError {
    fn code(&self) -> String {
        match self {
            LlmError::NoApiKey => "LLM_NO_API_KEY".into(),
            LlmError::Timeout => "LLM_TIMEOUT".into(),
            LlmError::Network => "LLM_NETWORK_ERROR".into(),
            LlmError::Http(code) => format!("LLM_HTTP_{code}"),
            LlmError::InvalidResponse => "LLM_INVALID_RESPONSE".into(),
        }
    }
}

#[derive(Deserialize)]
struct ChatCompletion {
    choices: Vec<Choice>,
}

#[derive(Deserialize)]
struct Choice {
    message: Message,
}

#[derive(Deserialize)]
struct Message {
    content: Option<String>,
}

#[tauri::command]
pub async fn generate_document(
    prompt: String,
    api_key: Option<String>,
    model: Option<String>,
    endpoint: Option<String>,
) -> Result<String, String> {
    let key = api_key
        .filter(|k| !k.trim().is_empty())
        .or_else(|| {
            std::env::var(API_KEY_ENV_VAR)
                .ok()
                .filter(|k| !k.trim().is_empty())
        })
        .ok_or_else(|| LlmError::NoApiKey.code())?;

    let url = endpoint.unwrap_or_else(|| DEFAULT_ENDPOINT.to_string());
    let model = model.unwrap_or_else(|| DEFAULT_MODEL.to_string());

    let client = reqwest::Client::new();
    let request = client
        .post(&url)
        .header("Authorization", format!("Bearer {key}"))
        .json(&serde_json::json!({
            "model": model,
            "messages": [
                {
                    "role": "system",
                    "content": "You are a professional government document writer. Respond only with a valid JSON object."
                },
                { "role": "user", "content": prompt }
            ],
            "temperature": 0.3,
            "response_format": { "type": "json_object" }
        }));

    let response = match tokio::time::timeout(REQUEST_TIMEOUT, request.send()).await {
        Ok(Ok(resp)) => resp,
        Ok(Err(_)) => return Err(LlmError::Network.code()),
        Err(_) => return Err(LlmError::Timeout.code()),
    };

    if !response.status().is_success() {
        return Err(LlmError::Http(response.status().as_u16()).code());
    }

    // send() 收到响应头即返回，读 body 必须单独包超时，否则服务端挂起 body 会无限等待。
    let completion: ChatCompletion = match tokio::time::timeout(REQUEST_TIMEOUT, response.json()).await {
        Ok(Ok(completion)) => completion,
        Ok(Err(_)) => return Err(LlmError::InvalidResponse.code()),
        Err(_) => return Err(LlmError::Timeout.code()),
    };

    completion
        .choices
        .into_iter()
        .next()
        .and_then(|c| c.message.content)
        .filter(|content| !content.trim().is_empty())
        .ok_or_else(|| LlmError::InvalidResponse.code())
}
